controllers/vm_controller.go

```diff
@@ -0,0 +1,140 @@
+package controllers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/hashicorp/go-multierror"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"kubevirt.io/ssp-operator/pkg/monitoring/metrics"
+)
+
+// +kubebuilder:rbac:groups=kubevirt.io,resources=virtualmachines,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=persistentvolumes,verbs=get;list;watch
+
+const vmControllerName = "vm-controller"
+
+type VmReconciler struct {
+	client client.Client
+	log    logr.Logger
+}
+
+func CreateVmController(mgr ctrl.Manager) (*VmReconciler, error) {
+	return newVmReconciler(mgr)
+}
+
+func (r *VmReconciler) Name() string {
+	return vmControllerName
+}
+
+func (r *VmReconciler) Start(_ context.Context, mgr ctrl.Manager) error {
+	return r.setupController(mgr)
+}
+
+func newVmReconciler(mgr ctrl.Manager) (*VmReconciler, error) {
+	logger := ctrl.Log.WithName("controllers").WithName("VirtualMachines")
+	reconciler := &VmReconciler{
+		client: mgr.GetClient(),
+		log:    logger,
+	}
+
+	return reconciler, nil
+}
+
+func (r *VmReconciler) setupController(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(vmControllerName).
+		For(&kubevirtv1.VirtualMachine{}).
+		Complete(r)
+}
+
+func (r *VmReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	r.log.Info("Starting vm reconciliation...", "request", req.String())
+	vm := kubevirtv1.VirtualMachine{}
+	if err := r.client.Get(ctx, req.NamespacedName, &vm); err != nil {
+		if errors.IsNotFound(err) {
+			// VM was deleted, so we can ignore it
+			r.log.Info("VM not found", "vm", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		r.log.Error(err, "Error getting VM", "vm", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	if err := r.setVmVolumesMetrics(&vm); err != nil {
+		r.log.Error(err, "Error setting vm volumes metrics", "vm", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("Finished vm reconciliation", "request", req.String())
+	return ctrl.Result{}, err
+}
+
+func (r *VmReconciler) setVmVolumesMetrics(vm *kubevirtv1.VirtualMachine) error {
+	var result error
+
+	if vm.Spec.Template.Spec.Volumes == nil {
+		return nil
+	}
+
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		volumeName := ""
+		if volume.DataVolume != nil {
+			volumeName = volume.DataVolume.Name
+		} else if volume.PersistentVolumeClaim != nil {
+			volumeName = volume.PersistentVolumeClaim.ClaimName
+		} else {
+			continue
+		}
+
+		pvc, err := r.getPVC(vm, volumeName)
+		if err != nil {
+			result = multierror.Append(result, err)
```

Comment:
I think it would be better to return the error and end the loop early, because if this function returns an error, the reconciliation will be triggered again.

Reply:
updated